protocol/communities/community_encryption_key_action.go

```diff
@@ -0,0 +1,105 @@
+		ChannelKeysActions: *evaluateChannelLevelEncryptionKeyActions(origin, modified, changes),
+	}
+	return result
+}
+
+func evaluateCommunityLevelEncryptionKeyAction(origin, modified *Community, changes *CommunityChanges) *EncryptionKeyAction {
+	originBecomeMemberPermissions := origin.TokenPermissionsByType(protobuf.CommunityTokenPermission_BECOME_MEMBER)
+	modifiedBecomeMemberPermissions := modified.TokenPermissionsByType(protobuf.CommunityTokenPermission_BECOME_MEMBER)
+
+	return evaluateEncryptionKeyAction(originBecomeMemberPermissions, modifiedBecomeMemberPermissions, changes.MembersAdded, changes.MembersRemoved)
+}
+
+func evaluateChannelLevelEncryptionKeyActions(origin, modified *Community, changes *CommunityChanges) *map[string]EncryptionKeyAction {
+	result := make(map[string]EncryptionKeyAction)
+
+	for chatID := range modified.config.CommunityDescription.Chats {
+		originChannelViewOnlyPermissions := origin.ChannelTokenPermissionsByType(chatID, protobuf.CommunityTokenPermission_CAN_VIEW_CHANNEL)
+		originChannelViewAndPostPermissions := origin.ChannelTokenPermissionsByType(chatID, protobuf.CommunityTokenPermission_CAN_VIEW_AND_POST_CHANNEL)
+		originChannelPermissions := append(originChannelViewOnlyPermissions, originChannelViewAndPostPermissions...)
+
+		modifiedChannelViewOnlyPermissions := modified.ChannelTokenPermissionsByType(chatID, protobuf.CommunityTokenPermission_CAN_VIEW_CHANNEL)
+		modifiedChannelViewAndPostPermissions := modified.ChannelTokenPermissionsByType(chatID, protobuf.CommunityTokenPermission_CAN_VIEW_AND_POST_CHANNEL)
+		modifiedChannelPermissions := append(modifiedChannelViewOnlyPermissions, modifiedChannelViewAndPostPermissions...)
+
+		membersAdded := make(map[string]*protobuf.CommunityMember)
+		membersRemoved := make(map[string]*protobuf.CommunityMember)
+
+		chatChanges, ok := changes.ChatsModified[chatID]
+		if ok {
+			membersAdded = chatChanges.MembersAdded
+			membersRemoved = chatChanges.MembersRemoved
+		}
+
+		result[chatID] = *evaluateEncryptionKeyAction(originChannelPermissions, modifiedChannelPermissions, membersAdded, membersRemoved)
+	}
+
+	return &result
+}
+
+func evaluateEncryptionKeyAction(originPermissions, modifiedPermissions []*protobuf.CommunityTokenPermission, membersAdded, membersRemoved map[string]*protobuf.CommunityMember) *EncryptionKeyAction {
+	result := &EncryptionKeyAction{
+		ActionType: EncryptionKeyNone,
+		Members:    map[string]*protobuf.CommunityMember{},
+	}
+
+	// permission was just added
+	if len(modifiedPermissions) > 0 && len(originPermissions) == 0 {
+		result.ActionType = EncryptionKeyAdd
+		return result
+	}
+
+	// permission was just removed
+	if len(modifiedPermissions) == 0 && len(originPermissions) > 0 {
+		result.ActionType = EncryptionKeyRemove
+		return result
+	}
+
+	// open community/channel does not require any actions
+	if len(modifiedPermissions) == 0 {
+		return result
+	}
+
+	if len(membersRemoved) > 0 {
+		result.ActionType = EncryptionKeyRekey
```

Comment:
Is there a possibility that a permission would be modified at the same time as a member would be removed? If so, maybe we should move this condition at the top, since it seems to have more impact than `EncryptionKeyAdd` for example.

Reply:
This utility determines if:
a) permission was just added (there was no permission before) - we send new key to everyone
b) permission was just removed (there was permission before, there is none anymore) - we "remove" the key (what we really do we simply stop encryption)
c) members changes - either rekey when any member was removed or send existing key to new members

If permission is modified at the same time as a member is removed it doesn't matter. All what matters are members changes, e.i. members will be removed or added depending if they satisfied the modified permission.